src/utils/editor/source-editor.js

```diff
@@ -18,40 +20,57 @@ require("codemirror/addon/search/searchcursor");
 // while shifting to a line which was initially out of view.
 const MAX_VERTICAL_OFFSET = 3;
 
+type SourceEditorOpts = {
+  enableCodeFolding: boolean,
+  extraKeys: Object,
+  gutters: string[],
+  lineNumbers: boolean,
+  lineWrapping: boolean,
+  matchBrackets: boolean,
+  mode: string,
+  readOnly: boolean,
+  showAnnotationRuler: boolean,
+  theme: string,
+  value: string,
+};
+
 class SourceEditor {
-  constructor(opts) {
+  opts: SourceEditorOpts;
+  editor: any;
+
+  constructor(opts: SourceEditorOpts) {
     this.opts = opts;
   }
 
-  appendToLocalElement(node) {
+  appendToLocalElement(node: any) {
     this.editor = CodeMirror(node, this.opts);
   }
 
   destroy() {
     // No need to do anything.
   }
 
-  get codeMirror() {
+  get codeMirror(): any {
     return this.editor;
   }
 
-  setText(str) {
+  setText(str: string) {
     this.editor.setValue(str);
   }
 
   getText() {
     return this.editor.getValue();
   }
 
-  setMode(value) {
+  setMode(value: Object) {
```

Comment:
Good call @clarkbw 

looks like in some cases it's a string

https://github.com/jasonLaster/debugger.html/blob/sparse2/src/utils/source.js#L89-L101

perhaps we add a mode type in `src/types` and use it in both `utils/source` and here

```js
let type Mode = String | {
  name: String,
  typescript?: String
  base?: {
    name: String,
    typescript: String
  }
}
```

frankly the typescript mode seems weird here. We should look at it again :)